src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,144 @@
+        },
+        () -> {});
+  }
+
+  // Idle shooter command; for default command purposes
+  public Command shooterIdle() {
+    return run(
+        () -> {
+          setShooterSpeeds(Constants.ShooterConstants.SHOOTER_IDLE_SPEEDS);
+        });
+  }
+
+  // Checks if shooter is ready.
+  public boolean isShooterReady(double distance) {
+    double target = calculateShooterSpeeds(distance)[0];
+    double minSpeedTolerance = target * (1 - Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+    double maxSpeedTolerance = target * (1 + Constants.ShooterConstants.SHOOTER_SPEED_TOLERANCE);
+
+    double vel = shooterTopMotor.getEncoder().getVelocity();
+    return (minSpeedTolerance < vel && vel < maxSpeedTolerance);
+  }
+
+  // Speed in RPM. Top is index 0, bottom is index 1.
+  public void setShooterSpeeds(double[] speeds) {
+    shooterTopMotor.getPIDController().setReference(speeds[0], CANSparkBase.ControlType.kVelocity);
+    shooterBottomMotor
+        .getPIDController()
+        .setReference(speeds[1], CANSparkBase.ControlType.kVelocity);
+  }
+
+  // Top is index 0, bottom is index 1.
+  public static double[] calculateShooterSpeeds(double distance) {
+    double[] speeds = {0, 0};
+    double distanceProportion;
+
+    // If distance below minimum, set speed to minimum.
+    if (distance < speakerConfigList[0].getDistance()) {
+      speeds[0] = speakerConfigList[0].getSpeedTop();
+      speeds[1] = speakerConfigList[0].getSpeedBottom();
+    } else {
+      // A linear search which determines which points the input distance falls between. May be
+      // converted to a binary search if there are many points
+      for (int i = 0; i < speakerConfigList.length; i++) {
+        // If distance above maximum, set speed to maximum.
+        if (i == speakerConfigList.length - 1) {
+          speeds[0] = speakerConfigList[i].getSpeedTop();
+          speeds[1] = speakerConfigList[i].getSpeedBottom();
+          break;
+        } else if (distance >= speakerConfigList[i].getDistance()
+            && distance < speakerConfigList[i + 1].getDistance()) {
+          // Math to linearly interpolate the speed.
+          distanceProportion =
+              (distance - speakerConfigList[i].getDistance())
+                  / (speakerConfigList[i + 1].getDistance() - speakerConfigList[i].getDistance());
+          speeds[0] =
+              (distanceProportion
+                      * (speakerConfigList[i + 1].getSpeedTop()
+                          - speakerConfigList[i].getSpeedTop()))
+                  + speakerConfigList[i].getSpeedTop();
+          speeds[1] =
+              (distanceProportion
+                      * (speakerConfigList[i + 1].getSpeedBottom()
+                          - speakerConfigList[i].getSpeedBottom()))
+                  + speakerConfigList[i].getSpeedBottom();
+          break;
+        }
+      }
+    }
+    return speeds;
+  }
+
+  private static class SpeakerConfig {
+    private double distance;
```

Comment:
It would be useful to have a unit attached to these fields, e.g. if the distance is in feet `distanceInFeet` and if the speed is in rotations per minute `speedTopRPM`.

Reply:
Resolved in 5967b58